File: Data Structures & Algorithms/products-of-array-discluding-self/submission-1.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {

        if (nums.size() == 0) {
            vector<int> out;
            return out;
        }
        
        if (nums.size() == 1) {
            vector<int> out;
            out.push_back(nums[0]);
            return out;
        }

        vector<int> prefix(nums.size(), 0);
        vector<int> postfix(nums.size(), 0);

        int prod = 1;

        for (int i = 0; i < nums.size(); i++) {
            if (nums[i] == 0) {
                break;
            }

            prod *= nums[i];
            prefix[i] = prod;
        }

        prod = 1;

        for (int i = nums.size() - 1; i >= 0; i--) {
            if (nums[i] == 0) {
                break;
            }

            prod *= nums[i];
            postfix[i] = prod;
        }

        vector<int> out(nums.size(), 0);
        
        out[0] = postfix[1];
        out[nums.size() - 1] = prefix[nums.size() - 2];

        for (int i = 1; i < nums.size() - 1; i++) {
            out[i] = prefix[i - 1] * postfix[i + 1];
        }

        return out;
    }
};

File: Data Structures & Algorithms/binary-search/submission-1.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        l = 0
        r = len(nums) - 1

        while l < r:

            m = l + (r - l) // 2
            nm = nums[m]

            if nm < target:
                l = m + 1
            elif nm > target:
                r = m - 1
            else:
                return m
        
        if nums[l] == target:
            return l
        
        return -1